Support submission of base16-encoded CBOR transactions (raw or wrapped as cardano-cli's text-envelope)

KtorZ opened the pull request:
- :round_pushpin: **Make parsing transactions from JSON more flexible: allow raw CBOR-encoded transactions.**
    This also supports cardano-cli's format out of the box, so one can really take any transaction coming from cardano-cli and push it through Hydra.

- :round_pushpin: **Cover new JSON parsing through simple unit tests.**
    The transaction is a well-formed Cardano transaction, created using the cardano-cli.

- :round_pushpin: **Derive text-envelope's type from Tx's type.**

- :round_pushpin: **Document / specify new client input API for transaction submission.**

KtorZ committed:
Make parsing transactions from JSON more flexible: allow raw CBOR-encoded transactions.

  This also supports cardano-cli's format out of the box, so one can really take any transaction coming from cardano-cli and push it through Hydra.

KtorZ committed:
Cover new JSON parsing through simple unit tests.

  The transaction is a well-formed Cardano transaction, created using the cardano-cli.

KtorZ committed:
Derive text-envelope's type from Tx's type.

KtorZ committed:
Document / specify new client input API for transaction submission.

KtorZ commented on hydra-node/src/Hydra/Ledger/Cardano/Json.hs:
```diff
@@ -471,17 +471,52 @@ instance
+    --
+    -- (c) As base16 string representing a CBOR-serialized transaction, since
+    -- this is the most common medium of exchange used for transactions.
+    parseAsJSONObject value <|> parseAsBase16CBOR value
```
On second-thought, this would be better written as two alternatives between three possibilities, instead of doing two of them within `parseAsJSONObject`; And we may want to put the raw base16-CBOR first as it's likely going to be the most commonly used.

ch1bo replied:
Would prefer the three alternatives as three parse functions as well, yes. The documentation already lends itself to be split between these three functions :)

ch1bo commented on hydra-node/test/Hydra/ClientInputSpec.hs:
```diff
@@ -23,6 +24,33 @@ spec = parallel $ do
     settings
     (Proxy @(ReasonablySized (ClientInput Tx)))
 
+  describe "FromJSON (ValidatedTx era)" $ do
```
We do have roundtrip tests on serialization of cardano transactions in `CardanoSpec`. IMO these tests would fit there better than in here. After all `ClientInput` has nothing to do with cardano transactions.

KtorZ replied:
> After all `ClientInput` has nothing to do with cardano transactions.

I... kindly disagree :sweat_smile: ? The `ClientInput` is the only place where this serialization is used (`NewTx`)! The `CardanoSpec` regards more the internal representation of various Cardano types, whereas the `ClientInputSpec` relates to our user-facing interface; There are round-trip tests just above but since the `ToJSON` instance will still produce JSON in the "detailed" form, they are no good for testing the alternative parsers.

ch1bo reviewed (approved):
Thanks for the contribution. Too bad we have not a feature about it on our roadmap :sweat_smile:

May I ask a quick note about these changes in the `CHANGELOG.md` as well?

Would be interesting to do a full client-side `hydraw` now!

KtorZ committed:
Split JSON parsing in three alternatives

  Better captures what's going on in the implementation, makes it easier to follow.

KtorZ committed:
Mention new base16&CBOR / Text-envelope support for 'NewTx' in CHANGELOG.
